File: backend/app/schemas/repository_search.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from app.schemas.test_case_tag import TestCaseTagCategory


class RepositorySearchTag(BaseModel):
    id: int
    category: TestCaseTagCategory
    name: str


class RepositorySuiteTag(RepositorySearchTag):
    test_case_count: int


class RepositoryTestCaseResult(BaseModel):
    type: Literal["test_case"] = "test_case"
    id: int
    label: str
    code: str
    title: str
    suite_id: int
    suite_name: str
    status: str
    tags: list[RepositorySearchTag] = Field(default_factory=list)
    business_area: RepositorySearchTag | None
    application_domain: RepositorySearchTag | None
    object_type: RepositorySearchTag | None


class RepositoryTestSuiteResult(BaseModel):
    type: Literal["test_suite"] = "test_suite"
    id: int
    label: str
    group_ids: list[int] = Field(default_factory=list)
    test_case_count: int
    is_active: bool
    tags: list[RepositorySuiteTag] = Field(default_factory=list)


class RepositorySuiteGroupResult(BaseModel):
    type: Literal["suite_group"] = "suite_group"
    id: int
    label: str
    parent_ids: list[int] = Field(default_factory=list)
    child_ids: list[int] = Field(default_factory=list)
    test_case_count: int
    tags: list[RepositorySuiteTag] = Field(default_factory=list)


RepositorySearchItem = Annotated[
    RepositoryTestCaseResult
    | RepositoryTestSuiteResult
    | RepositorySuiteGroupResult,
    Field(discriminator="type"),
]


class RepositorySearchResponse(BaseModel):
    query: str
    items: list[RepositorySearchItem]

File: frontend/src/api/repositorySearch.ts
import { request, type TestCaseStatus, type TestCaseTagCategory } from "./client";

export type RepositorySearchTag = {
  id: number;
  category: TestCaseTagCategory;
  name: string;
};

export type RepositorySuiteTag = RepositorySearchTag & {
  test_case_count: number;
};

export type RepositoryTestCaseResult = {
  type: "test_case";
  id: number;
  label: string;
  code: string;
  title: string;
  suite_id: number;
  suite_name: string;
  status: TestCaseStatus;
  business_area: RepositorySearchTag | null;
  application_domain: RepositorySearchTag | null;
  object_type: RepositorySearchTag | null;
  tags: RepositorySearchTag[];
};

export type RepositoryTestSuiteResult = {
  type: "test_suite";
  id: number;
  label: string;
  group_ids: number[];
  test_case_count: number;
  is_active: boolean;
  tags: RepositorySuiteTag[];
};

export type RepositorySuiteGroupResult = {
  type: "suite_group";
  id: number;
  label: string;
  parent_ids: number[];
  child_ids: number[];
  test_case_count: number;
  tags: RepositorySuiteTag[];
};

export type RepositorySearchItem = RepositoryTestCaseResult | RepositoryTestSuiteResult | RepositorySuiteGroupResult;
export type RepositorySearchType = "all" | "suites" | "groups" | "cases";

export type RepositorySearchParams = {
  query?: string;
  type?: RepositorySearchType;
  limit?: number;
  businessAreaIds?: number[];
  applicationDomainIds?: number[];
  objectTypeIds?: number[];
};

export type RepositorySearchResponse = {
  query: string;
  items: RepositorySearchItem[];
};

export function searchRepository(params: RepositorySearchParams) {
  const searchParams = new URLSearchParams();
  const query = params.query?.trim();
  if (query) searchParams.set("q", query);
  if (params.type === "suites") searchParams.set("types", "test_suite");
  if (params.type === "cases") searchParams.set("types", "test_case");
  if (params.type === "groups") searchParams.set("types", "suite_group");
  if (params.limit !== undefined) searchParams.set("limit", String(params.limit));
  for (const id of params.businessAreaIds ?? []) searchParams.append("business_area_id", String(id));
  for (const id of params.applicationDomainIds ?? []) searchParams.append("application_domain_id", String(id));
  for (const id of params.objectTypeIds ?? []) searchParams.append("object_type_id", String(id));

  return request<RepositorySearchResponse>(`/repository/search?${searchParams.toString()}`);
}
